## Dataset 2 - Predicting Salaries

In [1]:
import time
total_time = time.time()

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [3]:
adult = pd.read_csv('dataset2.csv')
print adult.head()
print adult.shape

         Id                                              Title  \
0  12612628                        Engineering Systems Analyst   
1  12612830                            Stress Engineer Glasgow   
2  12612844                   Modelling and simulation analyst   
3  12613049  Engineering Systems Analyst / Mathematical Mod...   
4  12613647         Pioneer, Miser Engineering Systems Analyst   

                                     FullDescription  \
0  Engineering Systems Analyst Dorking Surrey Sal...   
1  Stress Engineer Glasgow Salary **** to **** We...   
2  Mathematical Modeller / Simulation Analyst / O...   
3  Engineering Systems Analyst / Mathematical Mod...   
4  Pioneer, Miser  Engineering Systems Analyst Do...   

                         LocationRaw LocationNormalized ContractType  \
0            Dorking, Surrey, Surrey            Dorking          NaN   
1        Glasgow, Scotland, Scotland            Glasgow          NaN   
2  Hampshire, South East, South East          Hamp

In [4]:
X = adult[['Id', 'Title', 'FullDescription', 'LocationRaw', 'LocationNormalized', 'ContractType', 'ContractTime', 'Company', 'Category', 'SourceName']]
y = adult['SalaryNormalized']
df = df = pd.concat([X,y],axis=1)
df.head()

Id                                              Title  \
0  12612628                        Engineering Systems Analyst   
1  12612830                            Stress Engineer Glasgow   
2  12612844                   Modelling and simulation analyst   
3  12613049  Engineering Systems Analyst / Mathematical Mod...   
4  12613647         Pioneer, Miser Engineering Systems Analyst   

                                     FullDescription  \
0  Engineering Systems Analyst Dorking Surrey Sal...   
1  Stress Engineer Glasgow Salary **** to **** We...   
2  Mathematical Modeller / Simulation Analyst / O...   
3  Engineering Systems Analyst / Mathematical Mod...   
4  Pioneer, Miser  Engineering Systems Analyst Do...   

                         LocationRaw LocationNormalized ContractType  \
0            Dorking, Surrey, Surrey            Dorking          NaN   
1        Glasgow, Scotland, Scotland            Glasgow          NaN   
2  Hampshire, South East, South East          Hampshire          NaN   
3     Surrey, South East, South East             Surrey          NaN   
4     Surrey, South East, South East             Surrey          NaN   

  ContractTime                       Company          Category  \
0    permanent  Gregory Martin International  Engineering Jobs   
1    permanent  Gregory Martin International  Engineering Jobs   
2    permanent  Gregory Martin International  Engineering Jobs   
3    permanent  Gregory Martin International  Engineering Jobs   
4    permanent  Gregory Martin International  Engineering Jobs   

         SourceName  SalaryNormalized  
0  cv-library.co.uk             25000  
1  cv-library.co.uk             30000  
2  cv-library.co.uk             30000  
3  cv-library.co.uk             27500  
4  cv-library.co.uk             25000

In [5]:
for col in df:
    print col,df[col].isnull().sum()

Id 0
Title 1
FullDescription 0
LocationRaw 0
LocationNormalized 0
ContractType 6444
ContractTime 5263
Company 4049
Category 0
SourceName 0
SalaryNormalized 0


In [6]:
#na = df.dropna(inplace=True,axis=0)

In [7]:
for col in df:
    print col,df[col].isnull().sum()

Id 0
Title 1
FullDescription 0
LocationRaw 0
LocationNormalized 0
ContractType 6444
ContractTime 5263
Company 4049
Category 0
SourceName 0
SalaryNormalized 0


### Linear Regression


#### basic linear regression with only basic variables

In [8]:
def LinReg(num,X_dummies,yvar):
    import time
    start_time = time.time()
    Xdf = X_dummies
    # split data
    from sklearn.model_selection import train_test_split
    X_train,X_test,y_train,y_test = train_test_split(Xdf,yvar,test_size=0.5,random_state=1234)
    # scale data
    from sklearn.preprocessing import StandardScaler
    scaler = StandardScaler()
    X_train = pd.DataFrame(scaler.fit_transform(X_train.astype("float64")),columns=X_train.columns)
    X_test = scaler.transform(X_test.astype("float64"))
    # create
    from sklearn.linear_model import LinearRegression as LR
    lm = LR()
    lm.fit(X_train,y_train)
    pred = lm.predict(X_train)
    from sklearn.metrics import mean_squared_error as mse
    print 'RMSE:',(mse(y_test,pred))**.5
    print 'time:',time.time()-start_time

In [9]:
data = df
Xdum = pd.get_dummies(data[['Category','LocationNormalized','ContractTime']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 21181.4424725
time: 2.31900000572


In [10]:
data = df
Xdum = pd.get_dummies(data[['Category','LocationNormalized','ContractTime','ContractType']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 24585.7144135
time: 2.05999994278


In [11]:
data = df
Xdum = pd.get_dummies(data[['Category','ContractTime']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 17487.955297
time: 0.0490000247955


In [12]:
data = df
Xdum = pd.get_dummies(data[['SourceName']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 17727.7341769
time: 0.109999895096


In [13]:
data = df
Xdum = pd.get_dummies(data[['Category','Company']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 22411.4144717
time: 3.63700008392


In [14]:
data = df
Xdum = pd.get_dummies(data[['Company']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 25197.0501334
time: 2.78399991989


In [15]:
data = df
Xdum = pd.get_dummies(data[['Category','LocationNormalized']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 28146.2050072
time: 1.50499987602


In [16]:
data = df
Xdum = pd.get_dummies(data[['SourceName','LocationNormalized']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 21257.3240661
time: 1.76499986649


In [17]:
data = df
Xdum = pd.get_dummies(data[['LocationNormalized']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 22804.0920129
time: 1.31599998474


In [18]:
data = df
Xdum = pd.get_dummies(data[['SourceName','ContractTime']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 17798.0901632
time: 0.0629999637604


In [19]:
data = df
Xdum = pd.get_dummies(data[['SourceName','Category']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 17953.1762466
time: 0.0979998111725


In [20]:
data = df
Xdum = pd.get_dummies(data[['SourceName','Category','ContractTime']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 17974.0269877
time: 0.0999999046326


In [21]:
data = df
Xdum = pd.get_dummies(data[['Category','ContractTime','Company']]) 
y = data['SalaryNormalized']
lr1 = LinReg(1,Xdum,y)

RMSE: 21528.3907886
time: 2.51600003242


### ---> lowest RMSE with Linear Regression: 17488 with dummies for Category and ContractTime

### Running time

In [22]:
print time.time() - total_time

24.3000001907


## Linear Discriminant Analysis

In [23]:
def LDAModel(num,X_dummies,yvar):
    import time
    start_time = time.time()
    Xdf = X_dummies
    # split data
    from sklearn.model_selection import train_test_split
    X_train,X_test,y_train,y_test = train_test_split(Xdf,yvar,test_size=0.5,random_state=1234)
    # scale data
    from sklearn.preprocessing import StandardScaler
    scaler = StandardScaler()
    X_train = pd.DataFrame(scaler.fit_transform(X_train.astype("float64")),columns=X_train.columns)
    X_test = scaler.transform(X_test.astype("float64"))
    # create
    from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
    lda = LDA()
    #fit
    lda.fit(X_train,y_train)
    #test
    pred = lda.predict(X_test)
    #evaluate
    from sklearn.metrics import mean_squared_error as mse
    print 'RMSE:',(mse(y_test,pred))**.5
    print 'time:',time.time()-start_time

In [24]:
data = df
Xdum = pd.get_dummies(data[['Category','ContractTime']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 19403.3676435
time: 0.27799987793


C:\Users\russd\Anaconda2\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")


In [25]:
data = df
Xdum = pd.get_dummies(data[['SourceName','ContractTime']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 30675.7856562
time: 0.236999988556


In [26]:
data = df
Xdum = pd.get_dummies(data[['ContractTime']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 19145.8563773
time: 0.138000011444


In [27]:
data = df
Xdum = pd.get_dummies(data[['SourceName']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 18527.8137142
time: 0.244999885559


In [28]:
data = df
Xdum = pd.get_dummies(data[['Category','SourceName','LocationNormalized']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 30660.2362226
time: 4.25699996948


In [29]:
data = df
Xdum = pd.get_dummies(data[['Category','ContractTime','Company']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 17370.1717898
time: 4.90199995041


In [30]:
data = df
Xdum = pd.get_dummies(data[['SourceName']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 18527.8137142
time: 0.197000026703


In [31]:
data = df
Xdum = pd.get_dummies(data[['ContractTime']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 19145.8563773
time: 0.121999979019


In [32]:
data = df
Xdum = pd.get_dummies(data[['SourceName','ContractTime']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 30675.7856562
time: 0.183000087738


In [33]:
data = df
Xdum = pd.get_dummies(data[['SourceName','LocationNormalized','Category']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 30669.4085618
time: 5.86599993706


In [34]:
data = df
Xdum = pd.get_dummies(data[['SourceName','Category','ContractTime']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 17211.8911036
time: 0.384000062943


In [35]:
data = df
Xdum = pd.get_dummies(data[['Company','ContractTime','Category']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 30582.0243012
time: 8.45999979973


In [36]:
data = df
Xdum = pd.get_dummies(data[['ContractTime','Category']]) 
y = data['SalaryNormalized']
LDAModel(1,Xdum,y)

RMSE: 19403.3676435
time: 0.477999925613


### ---> lowest RSME with LDA: 17211 with Category, ContractTime, SourceName

## Quadratic Discriminant Analysis

In [37]:
def QDAModel(X_dummies,yvar):
    import time
    start_time = time.time()
    Xdf = X_dummies
    # split data
    from sklearn.model_selection import train_test_split
    X_train,X_test,y_train,y_test = train_test_split(Xdf,yvar,test_size=0.5,random_state=1234)
    # scale data
    from sklearn.preprocessing import StandardScaler
    scaler = StandardScaler()
    X_train = pd.DataFrame(scaler.fit_transform(X_train.astype("float64")),columns=X_train.columns)
    X_test = scaler.transform(X_test.astype("float64"))
    # create
    from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis as QDA
    qda = QDA()
    #fit
    qda.fit(X_train,y_train)
    #test
    pred = qda.predict(X_test)
    #evaluate
    from sklearn.metrics import mean_squared_error as mse
    print 'RMSE:',(mse(y_test,pred))**.5
    print 'time:',time.time()-start_time

### ---> lowest RSME with QDA: 17212 with SourceName, Category, ContractTime

## K Nearest Neighbors

In [38]:
def KNNModel(num,X_dummies,yvar,neighbors):
    import time
    start_time = time.time()
    Xdf = X_dummies
    # split data
    from sklearn.model_selection import train_test_split
    X_train,X_test,y_train,y_test = train_test_split(Xdf,yvar,test_size=0.5,random_state=1234)
    # scale data
    from sklearn.preprocessing import StandardScaler
    scaler = StandardScaler()
    X_train = pd.DataFrame(scaler.fit_transform(X_train.astype("float64")),columns=X_train.columns)
    X_test = scaler.transform(X_test.astype("float64"))
    # create
    from sklearn.neighbors import KNeighborsClassifier as KNN
    #create
    knn = KNN(n_neighbors=neighbors)
    #fit
    knn.fit(X_train,y_train)
    #test
    pred = knn.predict(X_test)
    #evaluate
    from sklearn.metrics import mean_squared_error as mse
    print 'RMSE:',(mse(y_test,pred))**.5
    print 'time:',time.time()-start_time

In [39]:
data = df
Xdum = pd.get_dummies(data[['ContractTime','Category']]) 
y = data['SalaryNormalized']
KNNModel(1,Xdum,y,1)

RMSE: 24576.629702
time: 5.02399992943


In [40]:
data = df
Xdum = pd.get_dummies(data[['ContractTime','Category']]) 
y = data['SalaryNormalized']
KNNModel(1,Xdum,y,3)

RMSE: 17884.1592208
time: 4.86299991608


In [41]:
data = df
Xdum = pd.get_dummies(data[['ContractTime','Category']]) 
y = data['SalaryNormalized']
KNNModel(1,Xdum,y,12)

RMSE: 19860.417021
time: 4.75100016594


In [42]:
data = df
Xdum = pd.get_dummies(data[['ContractTime','Category','SourceName']]) 
y = data['SalaryNormalized']
KNNModel(1,Xdum,y,3)

RMSE: 17614.6264075
time: 15.5789999962


### ---> lowest RSME with KNN: 17614 with ContractTime, Category, SourceName

## Naive Bayes

In [43]:
def NBModel(num,X_dummies,yvar):
    import time
    start_time = time.time()
    Xdf = X_dummies
    # split data
    from sklearn.model_selection import train_test_split
    X_train,X_test,y_train,y_test = train_test_split(Xdf,yvar,test_size=0.5,random_state=1234)
    # scale data
    from sklearn.preprocessing import StandardScaler
    scaler = StandardScaler()
    X_train = pd.DataFrame(scaler.fit_transform(X_train.astype("float64")),columns=X_train.columns)
    X_test = scaler.transform(X_test.astype("float64"))
    # create
    from sklearn.naive_bayes import GaussianNB as GNB
    gnb = GNB()
    gnb.fit(X_train,y_train)
    pred = gnb.predict(X_test)
    #evaluate
    from sklearn.metrics import mean_squared_error as mse
    print 'RMSE:',(mse(y_test,pred))**.5
    print 'time:',time.time()-start_time

In [44]:
data = df
Xdum = pd.get_dummies(data[['ContractTime','Category','SourceName']]) 
y = data['SalaryNormalized']
NBModel(1,Xdum,y)

RMSE: 37358.5591818
time: 11.754999876


In [45]:
data = df
Xdum = pd.get_dummies(data[['SourceName']]) 
y = data['SalaryNormalized']
NBModel(1,Xdum,y)

RMSE: 42647.1966903
time: 9.125


### ---> lowest RSME with NB: 37358 with ContractTime Category, SourceName

## run time

In [46]:
print 'total run time:',time.time() - total_time

total run time: 102.992000103
